### PR TITLE
License manager changes to enable license feature

### PR DESCRIPTION
**Pull request by abhilashraju**
Tested: Tested by executing following curl requests 
1) /redfish/v1/Registries/License(Get)
2) /redfish/v1/Registries/License/License
3) /redfish/v1/LicenseService(Get)
4) /redfish/v1/LicenseService/Licenses

Change-Id: Ieeeb1ceb0ad40c5cba2de0d7a1d4583b74abd5f8

### CODE REVIEW

**Comment by abhilashraju** [comment]
Some of the comments in https://github.com/ibm-openbmc/bmcweb/pull/902 is already taken care in this PR

**Comment by abhilashraju** [comment]
> > Some of the comments in #902 is already taken care in this PR
> 
> It's difficult to see which all comments are addressed here, since there is no diff. Please mark them in #902 for the reviewers to have a re-look.

Done

**Review by baemyung — COMMENTED** [review]
Among the comments, I think at least `.privileges()` parts need to be changed.

File: redfish-core/lib/license_service.hpp [hunk 1]
```diff
@@ -0,0 +1,680 @@
...
+#include <array>
+#include <filesystem>
+#include <string>
+#include <string_view>
+#include <variant>
+namespace redfish
+{
+namespace messages
+{
+static void addMessageToErrorJson(nlohmann::json& target,
```
[the remark this answers is not in the document]
Comment (abhilashraju) [reply redfish-core/lib/license_service.hpp:21]: why do we do that? isn't it cohisive to think that license error message belong to license  and keep it together?. I agree your point of having license_error_message.hpp. But I would prefer that to be auto generated from the pythone script. Licensemanager.hpp can include that header to use the apis. I was planning to add design discussion point on the same. The python script should be enhansed . I would like it to be run during build time. No need for checking in those headers. Actually Ed seems to have same opinion here https://gerrit.openbmc.org/c/openbmc/bmcweb/+/70120

File: redfish-core/lib/license_service.hpp [hunk 2]
```diff
@@ -0,0 +1,680 @@
...
+            }
+
+            uint64_t expirationTime = 0;
+            const uint32_t* deviceNumPtr = nullptr;
+            const std::string* serialNumPtr = nullptr;
+            const std::string* licenseNamePtr = nullptr;
+            const std::string* licenseTypePtr = nullptr;
+            const std::string* authorizationTypePtr = nullptr;
+            bool available = false;
+            bool state = false;
```
[the remark this answers is not in the document]
Comment (abhilashraju) [reply redfish-core/lib/license_service.hpp:526]: Will do in different PR

**Comment by gtmills** [comment]
Let's address the comments that can be addressed in a day. I don't want to leave all of these for a future refactoring.

**Comment by abhilashraju** [comment]
> Let's address the comments that can be addressed in a day. I don't want to leave all of these for a future refactoring.

It will take more than a day to incorporate all pending comments. I have already fixed a lot of refactor request. If you can approve this PR , I can start working on follow up PR to address the pending comments. Ideally a rebase commit should not have much review comment , as long as it is rebased on top latest master and free from any compilation error ,warning and basic sanity testing issue. Any review comment on top that is suggestion for re-impliment the feature using latest APIs available. Which is not in the scope of a rebase PR.

**Review by raviteja-b — APPROVED** [review]
we have already decided in design meetings earlier that License service can't be upstreamed and this is downstream only commit for coming up releases .

I agree with @abhilashraju here since this is downstream only feature, don't see much worth to rework on this PR anymore, lets merge this PR

**Comment by gtmills** [comment]
> It will take more than a day to incorporate all pending comments. I have already fixed a lot of refactor request. If you can approve this PR , I can start working on follow up PR to address the pending comments. Ideally a rebase commit should not have much review comment , as long as it is rebased on top latest master and free from any compilation error ,warning and basic sanity testing issue. Any review comment on top that is suggestion for re-impliment the feature using latest APIs available. Which is not in the scope of a rebase PR.

Let's do any that are simple (take less than a day here). Then I am good with this going in.

File: redfish-core/lib/license_service.hpp [hunk 3]
```diff
@@ -0,0 +1,680 @@
...
+        {
+            std::string entryID = object.first.filename();
+            if (entryID.empty())
+            {
+                continue;
+            }
+            entriesArray.push_back({});
+            nlohmann::json& thisEntry = entriesArray.back();
+            thisEntry["@odata.id"] = "/redfish/v1/LicenseService/Licenses/" +
+                                     entryID;
```
[the remark this answers is not in the document]
Comment (abhilashraju) [reply redfish-core/lib/license_service.hpp]: done

**Comment by abhilashraju** [comment]
> > It will take more than a day to incorporate all pending comments. I have already fixed a lot of refactor request. If you can approve this PR , I can start working on follow up PR to address the pending comments. Ideally a rebase commit should not have much review comment , as long as it is rebased on top latest master and free from any compilation error ,warning and basic sanity testing issue. Any review comment on top that is suggestion for re-impliment the feature using latest APIs available. Which is not in the scope of a rebase PR.
> 
> Let's do any that are simple (take less than a day here). Then I am good with this going in.

privilege and setProperty is pending . Which will take more than a day. Will do those in next PR

**Review by cmjishnu — APPROVED** [review]

**Comment by gtmills** [comment]
> privilege and setProperty is pending . Which will take more than a day. Will do those in next PR

I really just think you should do it here. 
I am fine if @baemyung is though.

**Comment by baemyung** [comment]
> > privilege and setProperty is pending . Which will take more than a day. Will do those in next PR
> 
> I really just think you should do it here. I am fine if @baemyung is though.

It would be great if it can be done here.
But, to make a progress, I'm good with the current code.  (Perhaps, we can do the refactoring soon).

**Review by baemyung — APPROVED** [review]

**Pull request merged**